### PR TITLE
Automatic infra creation for experiments

### PR DESCRIPTION
**Pull request by michelav**
Using Terraform for provisioning and Ansible for configuration. 

**Items created**
- Tracking, Training and Server Nodes;
- Cloud SQL Postgres Database with its user
- Brand New VPC with firewall rules
- Artifact Repository (Docker Registry)
- MLFLow service account. It's attached to servers so they can execute API calls without worrying with permissions

**Extra Configurations**
- All Nodes
  - Installing Docker, Pip and Conda
  - Creating user MLFlow with sudo and docker permissions
- Training Node
  - Set up MLFLOW_TRACKING_URI
- Tracking Node
  - Create a systemd service to start MLFlow Server

Still working on it bc there are issues!

### CODE REVIEW

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/nodes.tf [hunk 1]
```diff
@@ -0,0 +1,108 @@
+resource "google_compute_instance" "training_node" {
+  name                      = "training-node"
+  machine_type              = "c2-standard-8"
+  tags                      = ["ssh", "web", "tf-created"]
```
Comment (anayden) [inline Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/nodes.tf]: I wonder why the node is tagged with `web`. Maybe `training` would be more appropriate?

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/nodes.tf [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  machine_type              = "g1-small"
+  tags                      = ["ssh", "tracking", "tf-created"]
+  zone                      = "southamerica-east1-a"
+  allow_stopping_for_update = true
+
+  boot_disk {
+    initialize_params {
+      image = "ubuntu-2004-focal-v20210105"
+      size  = 100
+      type  = "pd-ssd"
```
Comment (savagej) [inline Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/nodes.tf]: tracker doesn't store any data so can use smaller, cheaper storage

**Review by savagej — APPROVED** [review]
Don't think I can give any good feedback on this, looks good though

**Comment by savagej** [comment]
What about the DB/bucket changes we talked about yesterday actually? We want them only provisioned once

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/nodes.tf (hunk 1 above)
Comment (michelav) [reply Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/nodes.tf]: You're right. Actually, I've started tagging by access protocols (bc of firewall rules) and later by roles. Forgot to put 'training' role there.

Comment (michelav) [reply Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/nodes.tf]:
```suggestion
  tags                      = ["ssh", "training", "tf-created"]
```

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/data.tf [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+resource "google_sql_database" "database" {
+  name     = var.mlflow_db_name
+  instance = google_sql_database_instance.englab_db_instance.name
+}
+
+# resource "google_storage_bucket" "mlflow_data_bucket" {
+#   name                        = "mlflow-data-bucket"
+#   location                    = var.gcp_region
+#   uniform_bucket_level_access = true
+# }
```
Comment (michelav) [inline Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/data.tf]: @savagej  It was commented out, but I'm also removing the block which creates a bucket. So, we gotta keep it outside of TF workflow.

```suggestion

```

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/data.tf [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  name             = "englab-${random_id.db_suffix.hex}"
+  database_version = "POSTGRES_12"
+
+  /*  CAREFUL: Just uncomment line below if you know what you're doing!
+   *  "deletion_protection" providers security for your even if you destroy
+   *  or delete your data. If set to 'true' (default), Terraform won't delete
+   *  your instance during a destroy command. You must set "delete_protecttion"
+   *  to false before trying to destroy.
+  */
+  deletion_protection = false
```
Comment (michelav) [inline Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/data.tf:21]: @savagej With this commit, the database instance won't be destroyed even if we do `terraform destroy`. We gotta set `deletion_protection` to false first. That means: change the property, tf apply and after that tf destroy. 

```suggestion
#  deletion_protection = false
```

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/output.tf [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+output "registry_url" {
+  description = "Private Docker Registry URI"
+  value       = "${google_artifact_registry_repository.englab_repository.location}-docker.pkg.dev"
+}
+
+# output "storage_url" {
+#   description = "MLFLow Data Storage URL"
+#   value = google_storage_bucket.mlflow_data_bucket.url
+# }
+
```
Comment (michelav) [inline Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/output.tf]: We don't control buckets anymore.

```suggestion


```

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/nodes.tf (hunk 2 above)
Comment (michelav) [reply Lab1_Operationalizing_Pytorch_with_Mlflow/src/infra/prov/nodes.tf]: Base image has 10GB.  What about 15GB? So, if we somehow we need extra space during conf and installation of something we "won't be catch with a short blanket".

```suggestion
      size  = 15
      type  = "pd-ssd"
```

**Pull request merged**